Makefile

```diff
@@ -98,22 +105,44 @@ ${CACHEDIR}/.libraries_token.txt: cache
 .PHONY: lib-token
 lib-token: ${CACHEDIR}/.libraries_token.txt
 
-.PHONY: fetch-kernel
+.PHONY: apk-token
 apk-token:
 	chainctl auth login --audience apk.cgr.dev
 
-fetch-kernel: apk-token
-	$(eval KERNEL_PKG := $(shell curl -L --silent --output - --user user:$$(chainctl auth token --audience apk.cgr.dev) https://apk.cgr.dev/chainguard-private/$(ARCH)/APKINDEX.tar.gz | \
-		zcat | \
-		grep -a -A1 "^P:linux$$" | \
-		grep "^V:" | \
-		sort -V | \
-		tail -n1 | sed -e "s/^V://"))
-	curl -s -LSo /tmp/linux.apk --user user:$(shell chainctl auth token --audience apk.cgr.dev) https://apk.cgr.dev/chainguard-private/$(ARCH)/linux-$(KERNEL_PKG).apk
-	mkdir -p /tmp/kernel
-	tar -xf /tmp/linux.apk -C /tmp/kernel/ 2>/dev/null
-	export QEMU_KERNEL_IMAGE=/tmp/kernel/boot/vmlinuz
-	export MELANGE_OPTS="--runner=qemu"
+.PHONY: fetch-kernel
+fetch-kernel:
+	rm -rf kernel
+	$(MAKE) kernel/boot/vmlinuz
+
+kernel/APKINDEX.tar.gz:
+	$(MAKE) apk-token
+	mkdir -p kernel
+	curl -LS --silent -o $@.tmp \
+	  --user user:$$(chainctl auth token --audience apk.cgr.dev) \
+	  https://apk.cgr.dev/chainguard-private/$(ARCH)/APKINDEX.tar.gz
+	mv $@.tmp $@
+
+kernel/APKINDEX: kernel/APKINDEX.tar.gz
+	tar -x -C kernel -f $< $(notdir $@)
+	touch $@
+
+kernel/chosen: kernel/APKINDEX
+	grep -e '^P:' -e '^V:' $< | sed -n '/^P:linux$$/{n;p}' > \
+	  kernel/available
+	grep '^V:' kernel/available | sed 's/V://' | \
+	  sort -V | tail -n1 > kernel/chosen
```

Comment:
2 things here. when you redirect like this, you create the file and then it will be up to date even if it fails.
it'd be good to verify that there is content you expected and then move it into place.

```
out=$$(grep '^V:' kernel/available | sed 's/V://' | sort -V | tail -n1) && [ -n "$$out" ] && echo "$$out" > $@)

Reply:
something like that, then you it doesnt get updated unless its sane.